ta/pkcs11/src/entry.c

```diff
@@ -35,40 +36,38 @@ void TA_CloseSessionEntryPoint(void *session __unused)
 /*
  * Entry point for invocation command PKCS11_CMD_PING
  *
- * @ctrl - param memref[0] or NULL: expected NULL
+ * @ctrl - param memref[0]: unused as input argument
  * @in - param memref[1] or NULL: expected NULL
- * @out - param memref[2] or NULL
+ * @out - param memref[2] or NULL: output data buffer for TA version info
  *
- * Return a TEE_Result value
+ * Return a PKCS11_CKR_* value
  */
-static TEE_Result entry_ping(TEE_Param *ctrl, TEE_Param *in, TEE_Param *out)
+static uint32_t entry_ping(TEE_Param *ctrl, TEE_Param *in, TEE_Param *out)
 {
 	const uint32_t ver[] = {
 		PKCS11_TA_VERSION_MAJOR,
 		PKCS11_TA_VERSION_MINOR,
 		PKCS11_TA_VERSION_PATCH,
 	};
-	size_t size = 0;
 
-	if (ctrl || in)
-		return TEE_ERROR_BAD_PARAMETERS;
-
-	if (!out)
-		return TEE_SUCCESS;
-
-	size = out->memref.size;
-	out->memref.size = sizeof(ver);
-
-	if (size < sizeof(ver))
-		return TEE_ERROR_SHORT_BUFFER;
+	/* Ctrl param does not carry input data: expected but not used here */
+	if (!ctrl || ctrl->memref.size != sizeof(uint32_t) || in ||
```

Comment:
Why doesn't this function update `ctrl` with the desired `PKCS11_*` value? You are checking that it has the correct size.
I assume that `ctrl` will be used for more than just carrying a `PKCS11_*` value and that the expected layout of `ctrl` depends on the function being called. This means that `TA_InvokeCommandEntryPoint()` isn't suitable to interpret and update the content of `ctrl`.

Reply:
I wanted to make it simple: use a generic API for parameters and not test each parameter type inside each specific command handler.
Maybe a not that good idea.